contracts/rewards/emission_manager_library.cairo

```diff
@@ -0,0 +1,214 @@
+%lang starknet
+
+from starkware.cairo.common.alloc import alloc
+from starkware.cairo.common.memcpy import memcpy
+from starkware.cairo.common.cairo_builtins import HashBuiltin
+from starkware.starknet.common.syscalls import get_caller_address
+
+from openzeppelin.access.ownable import Ownable
+from onlydust.stream.default_implementation import stream
+from onlydust.stream.generic import generic
+
+from contracts.interfaces.i_rewards_controller import IRewardsController
+from contracts.types.rewards_data import RewardsDataTypes
+
+#
+# Events
+#
+
+@event
+func emission_admin_updated(old_admin : felt, new_admin : felt):
+end
+
+#
+# Storage
+#
+
+@storage_var
+func EmissionManager_emission_admins(reward : felt) -> (admin : felt):
+end
+
+@storage_var
+func EmissionManager_rewards_controller() -> (address : felt):
+end
+
+namespace EmissionManager:
+    # Authorization
+
+    func assert_only_emission_admin{
+        syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr
+    }(reward : felt):
+        let (emission_admin) = EmissionManager_emission_admins.read(reward)
+        with_attr error_message("Only emission admin"):
+            let (caller) = get_caller_address()
+            assert caller = emission_admin
+        end
+        return ()
+    end
+
+    # Externals
+
+    func initializer{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+        owner : felt, rewards_controller : felt
+    ):
+        Ownable.initializer(owner)
+        EmissionManager_rewards_controller.write(rewards_controller)
+
+        return ()
+    end
+
+    func configure_assets{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
```

Comment:
code seems a bit complicated here, we don't need three additional inner functions (`_get_rewards_from_config`, `_validate_emission_admins` and `_validate_emission_admin_wrapper`). I'm not 100% sure, but would such code work:

```cairo
func configure_assets{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
    config_len : felt, config : RewardsDataTypes.RewardsConfigInput*
):
    alloc_locals

    if config_len == 0:
        return ()
    
    let (input) = [config]
    assert_only_emission_admin(input.reward_address)

    configure_assets(config_len - 1, config + RewardsDataTypes.RewardsConfigInput.SIZE)
    return ()
end
```

Reply:
I agree. Thought there's a point in which one would have to call the reward_controller function with the entire config (not yet implemented) after validation, I'm not sure how I would be able to send it after validating.

Reply:
The logic is: send tx with config to `emission_manager` -> `emission_manager` acts as a guard -> `emission_manager` calls `reward_controller`.

Reply:
Oh, I remembered, I did all of that in order to reuse some code [here](https://github.com/aave-starknet-project/aave-starknet-periphery/pull/2/files#diff-3446b96e43f987593e12c9be36c6af86ad9b223b763181983cdb48ced9edd8f1R113).
Now I'm not sure it's worth it.

Reply:
oh ok I got, we can still define two different inner functions. it would be more readable

Reply:
Agreed, that removes the cairo-stream library.